certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -250,6 +259,121 @@ def _get_ranked_matches(self, target_name):
+            out.remove(r"*")
+        return out
+
+    def choose_redirect_vhost(self, target_name, port):
+        r"""Chooses a virtual host for redirect enhancement.
+
+        Filter all hosts for those that include listen port but do not include
+        ssl on. Of these, find the most matching one based on target_name.
+
+        .. todo:: This should maybe return list if no obvious answer
+            is presented.
+
+        .. todo:: The special name "$hostname" corresponds to the machine's
+            hostname. Currently we just ignore this.
+
+        :param str target_name: domain name
+        :param str port: port number or "\*" or ""
+        :returns: vhost associated with name
+        :rtype: :class:`~certbot_nginx.obj.VirtualHost`
+
+        """
+        vhost = None
+
+        matches = self._get_redirect_ranked_matches(target_name, port)
+        if not matches:
+            # No matches for this port. That's fine.
+            return None
+        elif matches[0]['rank'] in xrange(2, 6):
+            # Wildcard match - need to find the longest one
+            rank = matches[0]['rank']
+            wildcards = [x for x in matches if x['rank'] == rank]
+            vhost = max(wildcards, key=lambda x: len(x['name']))['vhost']
+        else:
+            vhost = matches[0]['vhost']
+
+        return vhost
+
+    def _get_redirect_ranked_matches(self, target_name, port):
+        r"""Gets a ranked list of plaintextish port-listening vhosts matching target_name
+
+        Filter all hosts for those that include listen port but do not include
+        ssl on. Rank by how well these match target_name.
+
+        :param str target_name: The name to match
+        :param str port: port number or "\*" or ""
+        :returns: list of dicts containing the vhost, the matching name, and
+            the numerical rank
+        :rtype: list
+
+        """
+        # Nginx chooses a matching server name for a request with precedence:
+        # 1. exact name match
+        # 2. longest wildcard name starting with *
+        # 3. longest wildcard name ending with *
+        # 4. first matching regex in order of appearance in the file
+        matches = []
+        all_vhosts = self.parser.get_vhosts()
+        def _port_matches(test_port, matching_port):
+            # test_port is a number, matching is a number or * or ""
+            if matching_port == r"\*":
+                return True
+            if matching_port == "":
+                return test_port == self.DEFAULT_PORT
+            return test_port == matching_port
+
+        def _vhost_matches(vhost, port):
+            found_matching_port = False
+            for addr in vhost.addrs:
+                if _port_matches(port, addr.get_port()) and addr.ssl == False:
+                    found_matching_port = True
+            # if port is DEFAULT_PORT, check to see if there are no listen directives at all
+            if port == self.DEFAULT_PORT and len(vhost.addrs) == 0:
```

Comment:
*Maybe* easier to read if you do something like
```
if len(vhost.addrs) == 0:
  if port ==...:
else:
  for addr in vhost.addrs
```
to make it clear these are two alternative paths.